modules/finality-verifier/src/lib.rs

```diff
@@ -175,32 +322,207 @@ pub mod pallet {
 		FailedToWriteHeader,
 		/// There are too many requests for the current window to handle.
 		TooManyRequests,
+		/// The header being imported is on a fork which is incompatible with the current chain.
+		///
+		/// This can happen if we try and import a finalized header at a lower height than our
+		/// current `best_finalized` header.
+		ConflictingFork,
+		/// The scheduled authority set change found in the header is unsupported by the pallet.
+		///
+		/// This is the case for non-standard (e.g forced) authority set changes.
+		UnsupportedScheduledChange,
+		/// The pallet has already been initialized.
+		AlreadyInitialized,
+		/// All pallet operations are halted.
+		Halted,
+	}
+
+	/// Import the given header to the pallet's storage.
+	///
+	/// This function will also check if the header schedules and enacts authority set changes,
+	/// updating the current authority set accordingly.
+	///
+	/// Note: This function assumes that the given header has already been proven to be valid and
+	/// finalized. Using this assumption it will write them to storage with minimal checks. That
+	/// means it's of great importance that this function *not* called with any headers whose
+	/// finality has not been checked, otherwise you risk bricking your bridge.
+	pub(crate) fn import_header<T: Config>(header: BridgedHeader<T>) -> Result<(), sp_runtime::DispatchError> {
+		let best_finalized = <ImportedHeaders<T>>::get(<BestFinalized<T>>::get()).expect(
+			"In order to reach this point the bridge must have been initialized. Therefore
+			`ImportedHeaders` must contain an entry for `BestFinalized`.",
```

Comment:
This is only one part of the proof. When we initialize bridge indeed we set `BestFinalized` and also `ImportedHeaders`, but we should include in the proof a statement that every time `BestFinalized` is being set `ImportedHeaders` is updated as well.